### PR TITLE
M680x fixes

### PR DESCRIPTION
**Pull request by mmuman**
- [x] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

Some pending fixes for the m680x plugin.

### CODE REVIEW

**Comment by trufae** [comment]
Please rebase, those tests seems to work in local but not in the ci, so i marked them as broken for now

File: libr/arch/p/m680x_cs/plugin.c [hunk 1]
```diff
@@ -515,6 +515,14 @@ static bool decode(RArchSession *as, RAnalOp *op, RAnalOpMask mask) {
 					insn->op_str[0]?" ": "", insn->op_str);
 			r_str_replace_in (op->mnemonic, strlen (op->mnemonic) + 1,
 				"ptr ", "", true);
+			// we don't really output operands in intel order but we can clean up
+			if (as->config->syntax == R_ARCH_SYNTAX_INTEL) {
```
Comment (trufae) [inline libr/arch/p/m680x_cs/plugin.c:519]: uhm, is this $ an immediate and # a load on the given address? because if its a load we can use [addr] instead

**Comment by mmuman** [comment]
I didn't touch the tests yet. Actually which syntax are they run with anyway?

File: libr/arch/p/m680x_cs/plugin.c (hunk 1 above)
Comment (mmuman) [reply libr/arch/p/m680x_cs/plugin.c:519]: Oh it's actually just like for 6502, # is immediate, and $ is for hex, which is the easy part.

I could go like replace all the "lda " and other insns that accept immediate with "lda [", then replace "[#" by "", then append the missing ] to those lines that still have the [ …

**Comment by mmuman** [comment]
I made a naive version of the syntax stripper, but there are probably many other insns involved and side effects, like with multiple operands…

File: libr/arch/p/m680x_cs/plugin.c (hunk 1 above)
Comment (trufae) [reply libr/arch/p/m680x_cs/plugin.c:520]: yep, that may work

**Comment by trufae** [comment]
it is ready to review/merge?

**Comment by mmuman** [comment]
I'd like to finish classifying opcodes at least.

**Comment by trufae** [comment]
Sure, no hurry!

File: libr/arch/p/m680x_cs/plugin.c [hunk 2]
```diff
@@ -515,6 +515,25 @@ static bool decode(RArchSession *as, RAnalOp *op, RAnalOpMask mask) {
 					insn->op_str[0]?" ": "", insn->op_str);
 			r_str_replace_in (op->mnemonic, strlen (op->mnemonic) + 1,
 				"ptr ", "", true);
+			// we don't really output operands in intel order but we can clean up
+			// XXX: this is not finished
+			if (as->config->syntax == R_ARCH_SYNTAX_INTEL) {
+				op->mnemonic = r_str_replace (op->mnemonic,
```
Comment (trufae) [inline libr/arch/p/m680x_cs/plugin.c:521]: probably better to move this logic into a separate function, so op-> is only referenced once and not 10 times

File: libr/arch/p/m680x_cs/plugin.c [hunk 3]
```diff
@@ -515,6 +515,25 @@ static bool decode(RArchSession *as, RAnalOp *op, RAnalOpMask mask) {
...
+					"lda ", "lda [", true);
+				op->mnemonic = r_str_replace (op->mnemonic,
+					"ldx ", "ldx [", true);
+				op->mnemonic = r_str_replace (op->mnemonic,
+					"and ", "and [", true);
+				op->mnemonic = r_str_replace (op->mnemonic,
+					"ora ", "ora [", true);
+				op->mnemonic = r_str_replace (op->mnemonic,
+					" [#", " ", true);
+				if (strstr(op->mnemonic, "[")) {
```
Comment (trufae) [inline libr/arch/p/m680x_cs/plugin.c:533]: missing space before (

File: libr/arch/p/m680x_cs/plugin.c [hunk 4]
```diff
@@ -525,13 +544,38 @@ static bool decode(RArchSession *as, RAnalOp *op, RAnalOpMask mask) {
...
-		"gpr	pc	.16	48	0\n"
-		"gpr	sp	.16	48	0\n"
-		"gpr	a0	.16	48	0\n"
-		"gpr	a1	.16	48	0\n";
+		"=SP    s\n"
+		"gpr	pc	.16	0	0\n" ///< M6800/1/2/3/9, M6301/9
+		"gpr	s	.16	2	0\n" ///< M6809/M6309 system stack (=sp on others)
+		"gpr	cc	.8	4	0\n" ///< M6800/1/2/3/9, M6301/9
+		"flg	C		.1	4.0	0\n"
+		"flg	V		.1	4.1	0\n" // At least 6805 lacks it? Are the others shifted??
```
Comment (trufae) [inline libr/arch/p/m680x_cs/plugin.c:552]: two tabs. should be one

**Review by trufae — CHANGES_REQUESTED** [review]

**Comment by mmuman** [comment]
busy busy, I'll try to have a look next week.

**Comment by trufae** [comment]
the next week has arrived, can i help you somehow?

File: libr/arch/p/m680x_cs/plugin.c (hunk 2 above)
Comment (mmuman) [reply libr/arch/p/m680x_cs/plugin.c:557]: Indeed, and it's easier on the indentation.

File: libr/arch/p/m680x_cs/plugin.c (hunk 3 above)
Comment (mmuman) [reply libr/arch/p/m680x_cs/plugin.c:569]: ok

File: libr/arch/p/m680x_cs/plugin.c (hunk 4 above)
Comment (mmuman) [reply libr/arch/p/m680x_cs/plugin.c:588]: I probably copied that from some other place… fixed.

**Comment by mmuman** [comment]
Added some opcode types, but we should probably check them one by one, maybe sort the known ones by type instead, and move them from the alpha-sorted list gradually? We can probably do that later, I thought it was needed for function detection but it seems it's only useful for syntax coloring.

**Comment by mmuman** [comment]
bleh, JSR is not relative, had a doubt at last minute :sweat_smile:

File: libr/arch/p/m680x_cs/plugin.c (hunk 1 above)
Comment (mmuman) [reply libr/arch/p/m680x_cs/plugin.c:604]: Still incomplete but seems to work for now.

**Comment by mmuman** [comment]
ouch, bset != brset, it's not a jump…

File: libr/arch/p/m680x_cs/plugin.c [hunk 5]
```diff
@@ -67,6 +67,21 @@ static inline csh cs_handle_for_session (RArchSession *as) {
 	return pd->cs_handle;
 }
 
+static inline char *filter_intel_syntax (char *mnemonic) {
```
Comment (trufae) [inline libr/arch/p/m680x_cs/plugin.c:70]: i think its better not to inline this
```suggestion
static char *filter_intel_syntax(char *mnemonic) {
```

**Review by trufae — CHANGES_REQUESTED** [review]
looks good to merge fter this small style fix. you can add more improvements in a separate pr

File: libr/arch/p/m680x_cs/plugin.c (hunk 5 above)
Comment (mmuman) [reply libr/arch/p/m680x_cs/plugin.c:70]: ok!

**Comment by mmuman** [comment]
rebased

**Pull request merged**